ntoskrnl/include/internal/i386/mm.h

```diff
@@ -128,21 +117,68 @@
 #define MI_WRITE_VALID_PPE MI_WRITE_VALID_PTE
 
-/* Convert an address to a corresponding PTE */
-#define MiAddressToPte(x) \
-    ((PMMPTE)(((((ULONG)(x)) >> 12) << 2) + PTE_BASE))
+/*  Translating virtual addresses to physical addresses
+        (See: "Intel� 64 and IA-32 Architectures Software Developer�s Manual
+              Volume 3A: System Programming Guide, Part 1, CHAPTER 4 PAGING")
+    Page directory (PD) and Page table (PT) definitions
+    Page directory entry (PDE) and Page table entry (PTE) definitions
+*/
+
+/* Maximum number of page directories pages */
+#ifndef _X86PAE_
+#define PD_COUNT 1        /* Only one page directory page */
+#else
+#define PD_COUNT (1 << 2) /* The two most significant bits in the VA */
+#endif
+
+/* The number of PTEs on one page of the PT */
+#define PTE_PER_PAGE (PAGE_SIZE / sizeof(MMPTE))
+
+/* The number of PDEs on one page of the PD */
+#define PDE_PER_PAGE (PAGE_SIZE / sizeof(MMPDE))
+
+/* Maximum number of PDEs */
+#define PDE_PER_SYSTEM (PD_COUNT * PDE_PER_PAGE)
+
+/* TODO: It seems this variable is not needed for x86 */
+#define PPE_PER_PAGE 1
+
+/* Maximum number of pages for 4 GB of virtual space */
+#define MI_MAX_PAGES ((1ull << 32) / PAGE_SIZE)
+
+/* Base addresses for page tables */
+#define PTE_BASE (ULONG_PTR)0xC0000000
+#define PTE_TOP  (ULONG_PTR)(PTE_BASE + (MI_MAX_PAGES * sizeof(MMPTE)) - 1)
+#define PTE_MASK (PTE_TOP - PTE_BASE)
+
+#define MI_SYSTEM_PTE_BASE (PVOID)MiAddressToPte(NULL)
+
+/* Base addreses for page directories */
+#define PDE_BASE (ULONG_PTR)MiAddressToPte(PTE_BASE)
+#define PDE_TOP  (ULONG_PTR)(PDE_BASE + (PDE_PER_SYSTEM * sizeof(MMPDE)) - 1)
+#define PDE_MASK (PDE_TOP - PDE_BASE)
+
+/* The size of the virtual memory area that is mapped using a single PDE */
+#ifndef PDE_MAPPED_VA
+#define PDE_MAPPED_VA  (PTE_PER_PAGE * PAGE_SIZE)
+#endif
+
+/* Maps the virtual address to the corresponding PTE */
+#define MiAddressToPte(Va) \
+    ((PMMPTE)(PTE_BASE + ((((ULONG_PTR)(Va)) / PAGE_SIZE) * sizeof(MMPTE))))
+
+/* Maps the virtual address to the corresponding PDE */
+#define MiAddressToPde(Va) \
+    ((PMMPDE)(PDE_BASE + ((MiAddressToPdeOffset(Va)) * sizeof(MMPTE))))
 
-/* Convert an address to a corresponding PDE */
-#define MiAddressToPde(x) \
-    ((PMMPDE)(((((ULONG)(x)) >> 22) << 2) + PDE_BASE))
+/* Takes the PTE index (for one PD page) from the virtual address */
+#define MiAddressToPteOffset(Va) \
+    ((((ULONG_PTR)(Va)) & (PDE_MAPPED_VA - 1)) / PAGE_SIZE)
 
-/* Convert an address to a corresponding PTE offset/index */
-#define MiAddressToPteOffset(x) \
-    ((((ULONG)(x)) << 10) >> 22)
+/* Takes the PDE offset (within all PDs pages) from the virtual address */
+#define MiAddressToPdeOffset(Va) (((ULONG_PTR)(Va)) / PDE_MAPPED_VA)
 
-/* Convert an address to a corresponding PDE offset/index */
-#define MiAddressToPdeOffset(x) \
-    (((ULONG)(x)) / (1024 * PAGE_SIZE))
+/* TODO: Free this variable (for offset from the pointer to the PDE) */
```

Comment:
I don't understand the comment. Do you mean "Get rid of this define" ?

Reply:
I want to use MiAddressToPdeOffset() wherever MiGetPdeOffset() is currently used. Thus freeing this name for use where the parameter is a pointer to PDE (MiGetPdeOffset(_Pde)) instead of MiAddressToPdeOffset (Va).